libparsec/crates/client/src/certificates/realm_keys_bundle.rs

```diff
@@ -0,0 +1,586 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use std::sync::Arc;
+
+use libparsec_client_connection::ConnectionError;
+use libparsec_platform_storage::certificates::{GetCertificateError, UpTo};
+use libparsec_types::prelude::*;
+
+use super::{store::CertificatesStoreReadGuard, CertificatesOps};
+
+#[derive(Debug)]
+enum ValidatedKey {
+    Valid {
+        key: SecretKey,
+        timestamp: DateTime,
+    },
+    /// Key has failed to decrypt the canary of the corresponding key rotation certificate.
+    /// As it name imply, the corrupted key should not be used. But we need to keep it
+    /// nevertheless to be able to generate the next keys bundle.
+    Corrupted {
+        corrupted_key_dont_use_me: SecretKey,
+    },
+}
+
+#[derive(Debug)]
+pub(super) struct RealmEncryptionInfo {
+    pub realm_id: VlobID,
+    // Not public field: use `last_valid_key()` & `key_from_index()` methods to access
+    // the keys (as this prevent us from accidentally using a corrupted key).
+    keys: Vec<ValidatedKey>,
+    // Once keys bundle decrypted, keys bundle access key is no longer needed.
+    // However it is still useful to keep it around given we need it to transmit
+    // it when sharing the workspace (and validating the keys bundle access fetched
+    // from server is cumbersome and error prone, so better to do it once !).
+    pub keys_bundle_access_key: SecretKey,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum KeyFromIndexError {
+    #[error("Key exists but is corrupted")]
+    CorruptedKey,
+    #[error("Key doesn't exist at this time")]
+    KeyNotFound,
+}
+
+impl RealmEncryptionInfo {
+    pub fn key_index(&self) -> IndexInt {
+        self.keys.len() as IndexInt
+    }
+
+    /// Return `None` if no valid key exists :(
+    pub fn last_valid_key(&self) -> Option<(&SecretKey, IndexInt)> {
+        self.keys
+            .iter()
+            .rev()
+            .enumerate()
+            .find_map(|(index, key)| match key {
+                ValidatedKey::Valid { key, .. } => Some((key, (index + 1) as IndexInt)),
+                ValidatedKey::Corrupted { .. } => None,
+            })
+    }
+
+    pub fn key_from_index(
+        &self,
+        key_index: IndexInt,
+        up_to: DateTime,
+    ) -> Result<&SecretKey, KeyFromIndexError> {
+        let key = (key_index as usize)
+            .checked_sub(1)
+            .map(|index| self.keys.get(index))
+            .flatten()
+            .ok_or(KeyFromIndexError::KeyNotFound)?;
+
+        match key {
+            ValidatedKey::Valid { key, timestamp } if *timestamp <= up_to => Ok(key),
+            ValidatedKey::Valid { .. } => Err(KeyFromIndexError::KeyNotFound),
+            ValidatedKey::Corrupted { .. } => Err(KeyFromIndexError::CorruptedKey),
+        }
+    }
```

Comment:
It is very strange that you `.enumerate()` after the `.rev()` `l.55-56` and you use the same index with `.get()` `l.70` This seems a bug introduced if you have more than `1` key.

Reply:
Nice catch !

Given I use the index in the array as the key index (with a minus 1, guess this is the only time we can say lua was right :trollface: ), I should obviously first enumerate, then reverse the list !